Migrate CSS/Style value types off of CSSUnitType and onto type specific unit enums

weinig opened the pull request:
#### bad39b1ccef9ee059c6f045f2efd81796fddb896
<pre>
Migrate CSS/Style value types off of CSSUnitType and onto type specific unit enums
<a href="https://bugs.webkit.org/show_bug.cgi?id=285269">https://bugs.webkit.org/show_bug.cgi?id=285269</a>

Reviewed by Darin Adler.

Replaces use of CSSUnitType in the CSS/Style value types with
type specific unit enums (e.g. LengthUnit, AngleUnit, etc.)

This presents a good opportunity to start moving functionality
out of CSSPrimitiveValue.h/cpp. Here we move number and unit
serialization and length resolution out.

The use of type specific unit enums also means we can make
some initialization patterns a bit cleaner by avoid nested
type initialization for numeric types in most cases.

With unit types in place, we also now have literal syntax
for any CSS numeric value. For example, `50px` is:

    50_css_px

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSGradientValue.cpp:
* Source/WebCore/css/CSSPrimitiveValue.cpp:
* Source/WebCore/css/CSSPrimitiveValue.h:
* Source/WebCore/css/calc/CSSCalcTree+CalculationValue.cpp:
* Source/WebCore/css/calc/CSSCalcTree+ComputedStyleDependencies.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Serialization.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Simplification.cpp:
* Source/WebCore/css/calc/CSSCalcValue.cpp:
* Source/WebCore/css/parser/CSSParserFastPaths.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+AngleDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+AnglePercentageDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+CSSPrimitiveValueResolver.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+FlexDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+FrequencyDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+IntegerDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+LengthDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+LengthPercentageDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+MetaConsumerDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+NumberDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+PercentageDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+ResolutionDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Shapes.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+TimeDefinitions.h:
* Source/WebCore/css/parser/SizesAttributeParser.cpp:
* Source/WebCore/css/parser/SizesAttributeParser.h:
* Source/WebCore/css/parser/SizesCalcParser.cpp:
* Source/WebCore/css/values/backgrounds/CSSBorderRadius.cpp:
* Source/WebCore/css/values/backgrounds/CSSBorderRadius.h:
* Source/WebCore/css/values/color/CSSColorConversion+Normalize.h:
* Source/WebCore/css/values/color/CSSColorConversion+ToTypedColor.h:
* Source/WebCore/css/values/color/CSSColorDescriptors.h:
* Source/WebCore/css/values/filter-effects/CSSFilterFunctionDescriptor.h:
* Source/WebCore/css/values/images/CSSGradient.cpp:
* Source/WebCore/css/values/primitives/CSSPosition.cpp:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericConcepts.h: Added.
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericRaw.h: Added.
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes+CSSValueCreation.h:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes+Canonicalization.cpp:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes+Canonicalization.h:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes+ComputedStyleDependencies.cpp:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes+ComputedStyleDependencies.h:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes+Serialization.cpp:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes+Serialization.h:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes+SymbolReplacement.h:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes.cpp: Removed.
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes.h:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericUnits.cpp: Added.
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericUnits.h: Added.
* Source/WebCore/css/values/shapes/CSSShapeFunction.cpp:
* Source/WebCore/editing/FontAttributeChanges.cpp:
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
* Source/WebCore/page/PrintContext.cpp:
* Source/WebCore/style/MatchedDeclarationsCache.cpp:
* Source/WebCore/style/StyleResolveForFont.cpp:
* Source/WebCore/style/values/images/StyleGradient.cpp:
* Source/WebCore/style/values/primitives/StyleLengthResolution.cpp: Added.
* Source/WebCore/style/values/primitives/StyleLengthResolution.h: Added.
* Source/WebCore/style/values/primitives/StylePrimitiveNumericConcepts.h: Added.
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Conversions.cpp:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Conversions.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Evaluation.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Logging.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes.h:

Canonical link: <a href="https://commits.webkit.org/288446@main">https://commits.webkit.org/288446@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1008bb922bb6463696578f0c0854f6ca2924e262

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/83399 "35 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/3016 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/37687 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/88475 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/34409 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/85487 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/3101 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/10971 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/64878 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/22627 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/86449 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/2261 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/75786 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/45162 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/2168 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/29989 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/33458 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/76360 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/73271 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/30717 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/89851 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/82418 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/10663 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/7690 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/73306 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/10888 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/71608 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/72536 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17955 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/16754 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/15487 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/1999 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/10618 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/16090 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/104838 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/10469 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25355 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/13940 "Built successfully") | | | 
| | [❌ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/12241 "Hash 1008bb92 for PR 38443 does not build (failure)") | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/style/values/StyleValueTypes.h:
```diff
@@ -264,16 +267,16 @@ template<typename T, size_t N> struct ToStyleMapping<CommaSeparatedVector<T, N>>
 
-// Constrained for `TreatAsNonConverting`.
-template<typename CSSType> requires (TreatAsNonConverting<CSSType>) struct ToStyle<CSSType> {
+// Constrained for `TreatAs`.
```
Oops, didn't mean to change this comment. Will fix before landing.

weinig commented on Source/WebCore/style/values/primitives/StylePrimitiveNumericConcepts.h:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2024 Samuel Weinig <sam@webkit.org>
+ * Copyright (C) 2025 Samuel Weinig <sam@webkit.org>
```
This is just git being funny. This is a totally new file, not a rename. Hence the date change.

darinadler commented on Source/WebCore/css/calc/CSSCalcTree+ComputedStyleDependencies.cpp:
```diff
@@ -49,10 +49,16 @@ static void collectComputedStyleDependencies(const Child& root, ComputedStyleDep
+            auto lengthUnit = CSS::toLengthUnit(root.unit);
+            if (!lengthUnit)
+                return;
+            CSS::collectComputedStyleDependencies(dependencies, *lengthUnit);
```
```suggestion
            if (auto lengthUnit = CSS::toLengthUnit(root.unit))
                CSS::collectComputedStyleDependencies(dependencies, *lengthUnit);
```

darinadler commented on Source/WebCore/css/values/primitives/CSSPrimitiveNumericUnits.h:
```diff
@@ -0,0 +1,1532 @@
+        consteval ValueLiteral<type> operator""_css_##name(long double value) { return { static_cast<double>(value) }; } \
+        consteval ValueLiteral<type> operator""_css_##name(unsigned long long value) { return { static_cast<double>(value) }; } \
+    } \
+\
```
I usually put something else on the last line, like maybe `//`, can't remember exactly. This style seems a little risky since it picks up the line after, too

darinadler commented on Source/WebCore/css/values/primitives/CSSPrimitiveNumericUnits.h:
```diff
@@ -0,0 +1,1532 @@
+        || lengthUnit == Lh
+        || lengthUnit == Cap
+        || lengthUnit == Ch
+        || lengthUnit == Ic;
```
Should these use `switch` instead?

weinig replied:
Yeah, that would be better. I copied them from CSSPrimitiveValue without thinking too much. Will fix.

weinig replied:
I really would love to write these at all, and instead have this done declaratively. Maybe use macros to define the units with additional metadata? Maybe one of these things:

```c++
#define FOR_EACH_LENGTH_UNIT(CASE)  \
    CASE(Px, px, CSSUnitType::CSS_PX, LengthUnitKind::None) \
    ...
    CASE(Em, em, CSSUnitType::CSS_EM, LengthUnitKind::FontRelative) \
```
 
For another round though.

darinadler commented on Source/WebCore/css/values/primitives/CSSPrimitiveNumericUnits.h:
```diff
@@ -0,0 +1,1532 @@
+// Allows unchecked casting DOWN from a composite unit type to one of the unit types it includes.
+//    `AngleUnit::Deg` == uncheckedUnitDowncast<AngleUnit>(AnglePercentageUnit::Deg)
+//    "undefined"      == uncheckedUnitDowncast<AngleUnit>(AnglePercentageUnit::Percentage)
+// NOTE: Use with cation. Only a debug assertion will check that the cast is valid.
```
Offering this goes against trend. We are doing more checking because of the security benefits. This should be here only if there is a really good reason it needs to be.

weinig replied:
I think I didn’t end up needing it once I added the switchOnUnit function so can remove it.

darinadler commented on Source/WebCore/css/values/primitives/CSSPrimitiveNumericUnits.h:
```diff
@@ -0,0 +1,1532 @@
+// Allows unchecked casting DOWN from a composite unit type to one of the unit types it includes.
+//    `AngleUnit::Deg` == uncheckedUnitDowncast<AngleUnit>(AnglePercentageUnit::Deg)
+//    "undefined"      == uncheckedUnitDowncast<AngleUnit>(AnglePercentageUnit::Percentage)
+// NOTE: Use with cation. Only a debug assertion will check that the cast is valid.
```
```suggestion
// NOTE: Use with caution. Only a debug assertion will check that the cast is valid.
```

darinadler commented on Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
```diff
@@ -3073,12 +3084,20 @@ void CanvasRenderingContext2DBase::setWordSpacing(const String& wordSpacing)
+    if (!rawLength)
+        return;
+
+    // FIXME: Ensure rawLength->unit is supported by `Style::computeUnzoomedNonCalcLengthDouble`.
```
Dows this shortcoming have test coverage?

weinig replied:
It’s an existing problem I am fixing in a follow up PR with tests.

darinadler commented on Source/WebCore/style/values/primitives/StyleLengthResolution.cpp:
```diff
@@ -0,0 +1,423 @@
+
+    case BoxAxis::Vertical:
+        return size.height();
+    }
```
```suggestion
    switch (mapAxisLogicalToPhysical(style->writingMode(), logicalAxis)) {
    case BoxAxis::Horizontal:
        return size.width();
    case BoxAxis::Vertical:
        return size.height();
    }
```

weinig committed:
Migrate CSS/Style value types off of CSSUnitType and onto type specific unit enums
https://bugs.webkit.org/show_bug.cgi?id=285269

Reviewed by Darin Adler.

Replaces use of CSSUnitType in the CSS/Style value types with
type specific unit enums (e.g. LengthUnit, AngleUnit, etc.)

This presents a good opportunity to start moving functionality
out of CSSPrimitiveValue.h/cpp. Here we move number and unit
serialization and length resolution out.

The use of type specific unit enums also means we can make
some initialization patterns a bit cleaner by avoid nested
type initialization for numeric types in most cases.

With unit types in place, we also now have literal syntax
for any CSS numeric value. For example, `50px` is:

    50_css_px

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSGradientValue.cpp:
* Source/WebCore/css/CSSPrimitiveValue.cpp:
* Source/WebCore/css/CSSPrimitiveValue.h:
* Source/WebCore/css/calc/CSSCalcTree+CalculationValue.cpp:
* Source/WebCore/css/calc/CSSCalcTree+ComputedStyleDependencies.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Serialization.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Simplification.cpp:
* Source/WebCore/css/calc/CSSCalcValue.cpp:
* Source/WebCore/css/parser/CSSParserFastPaths.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+AngleDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+AnglePercentageDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+CSSPrimitiveValueResolver.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+FlexDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+FrequencyDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+IntegerDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+LengthDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+LengthPercentageDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+MetaConsumerDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+NumberDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+PercentageDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+ResolutionDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Shapes.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+TimeDefinitions.h:
* Source/WebCore/css/parser/SizesAttributeParser.cpp:
* Source/WebCore/css/parser/SizesAttributeParser.h:
* Source/WebCore/css/parser/SizesCalcParser.cpp:
* Source/WebCore/css/values/backgrounds/CSSBorderRadius.cpp:
* Source/WebCore/css/values/backgrounds/CSSBorderRadius.h:
* Source/WebCore/css/values/color/CSSColorConversion+Normalize.h:
* Source/WebCore/css/values/color/CSSColorConversion+ToTypedColor.h:
* Source/WebCore/css/values/color/CSSColorDescriptors.h:
* Source/WebCore/css/values/filter-effects/CSSFilterFunctionDescriptor.h:
* Source/WebCore/css/values/images/CSSGradient.cpp:
* Source/WebCore/css/values/primitives/CSSPosition.cpp:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericConcepts.h: Added.
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericRaw.h: Added.
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes+CSSValueCreation.h:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes+Canonicalization.cpp:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes+Canonicalization.h:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes+ComputedStyleDependencies.cpp:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes+ComputedStyleDependencies.h:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes+Serialization.cpp:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes+Serialization.h:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes+SymbolReplacement.h:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes.cpp: Removed.
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes.h:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericUnits.cpp: Added.
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericUnits.h: Added.
* Source/WebCore/css/values/shapes/CSSShapeFunction.cpp:
* Source/WebCore/editing/FontAttributeChanges.cpp:
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
* Source/WebCore/page/PrintContext.cpp:
* Source/WebCore/style/MatchedDeclarationsCache.cpp:
* Source/WebCore/style/StyleResolveForFont.cpp:
* Source/WebCore/style/values/images/StyleGradient.cpp:
* Source/WebCore/style/values/primitives/StyleLengthResolution.cpp: Added.
* Source/WebCore/style/values/primitives/StyleLengthResolution.h: Added.
* Source/WebCore/style/values/primitives/StylePrimitiveNumericConcepts.h: Added.
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Conversions.cpp:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Conversions.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Evaluation.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Logging.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes.h:

Canonical link: https://commits.webkit.org/288446@main
